server/aesdsocket.c

```diff
@@ -138,121 +355,130 @@ int main(int argc, char** argv){
+	if ( rc != 0 ) {
+	    	free(timeStampthreadParams);
+		syslog(LOG_ERR,"Creation of thread %lu failed with code %d\n",timeStampThread, rc);
+	}
+	
 	/* Restarts accepting connections from new clients forever in a loop until SIGINT or SIGTERM is received */
 	while(1) {
 	
+		if(signal_caught){
+			break;
+		}
+	
 		/* Accepts incoming connection on port 9000 and stores incoming address*/
 		struct sockaddr_storage incomingAddr;
 	  	socklen_t addrSize = sizeof(incomingAddr);
 	    	int connFd = accept(sockfd, (struct sockaddr *)&incomingAddr, &addrSize);
 		if (connFd < 0){
-			syslog(LOG_ERR, "Error accepting connection on socket with error: %s", strerror(errno));
-			exit(-1);
+			if(signal_caught) {
+				break;
+			} else {
+				syslog(LOG_ERR, "Error accepting connection on socket with error: %s", strerror(errno));
+				exit(-1);
+			}
 		}
 		
-		/* Get the ip address of the connecting client */
-		char ipv4Addr[INET_ADDRSTRLEN];
-		struct sockaddr *incomingSockAddr = (struct sockaddr *)&incomingAddr;
-	    	void *binAddr = &(((struct sockaddr_in  *)incomingSockAddr)->sin_addr);
-	    	const char *res = inet_ntop(incomingAddr.ss_family, &binAddr, ipv4Addr, sizeof(ipv4Addr));
-	    	if (res == NULL){
-    			syslog(LOG_ERR, "Error converting client IP Address with error: %s", strerror(errno));
-			exit(1);
-	    	}
-		syslog(LOG_DEBUG, "Accepted connection from %s", ipv4Addr);
+		/* Create a new thread for the connection */
+		pthread_t connThread;
 		
-		/* Receive latest data packet from client into buffer */
-		int bytesRead  = 0;
-		char inBuf[BUFFERSIZE];
-		bool endStream = false;
+		/* Initialise threadParams data struct dynamically */
+		struct connThread *threadParams = (struct connThread*) malloc(sizeof(struct connThread));
 		
-		/* While data comes in through the stream */
-		while(!endStream){
-			bytesRead = recv(connFd, inBuf, sizeof(inBuf), 0);
-			if(bytesRead <= 0){
-				endStream = true;
-				break;
-			}
-			totalBytes += bytesRead;
-			
-			/* Append received data to file /var/tmp/aesdsocketdata */
-			int fd = open(FILE_PATH, O_RDWR | O_CREAT, S_IRWXU | S_IRWXG | S_IRWXO);
-			/* Error opening file: error printed and logged and program terminates with status 1 */
-			if (fd ==-1) 
-			{
-				syslog(LOG_ERR, "Error: (%s) while opening %s", strerror(errno), FILE_PATH);
-				exit(1);
-			}
-				
-			lseek(fd, 0, SEEK_END);
-			int bytesWritten = write(fd, &inBuf, bytesRead);
-			/* Error writing to file */
-			if (bytesWritten==-1){
-				syslog(LOG_ERR, "Error: (%s) while writing to %s", strerror(errno), FILE_PATH);
-				exit(1);
-			} 
-			/* Incomplete write to file */
-			else if (bytesWritten < bytesRead){
-				syslog(LOG_ERR, "Error: Incomplete write to %s, expected: %d bytes but wrote: %d bytes", FILE_PATH, bytesRead, bytesWritten);
-			}
-		        close(fd);
+		/* Populate data struct with arguments to the function */
+		threadParams->connFd = connFd;
+		threadParams->thread = &connThread;		
```

Comment:
I don't think you want to use &connThread here, as it will result in every thread using the same storage location on the stack for its thread ID.  I think this is ultimately the reason for your valgrind possibly lost message.